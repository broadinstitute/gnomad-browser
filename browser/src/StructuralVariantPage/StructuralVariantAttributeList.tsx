import React from 'react'

import { Badge } from '@gnomad/ui'

import AttributeList, { AttributeListItem } from '../AttributeList'
import InfoButton from '../help/InfoButton'
import Link from '../Link'
import { svTypeLabels } from '../StructuralVariantList/structuralVariantTypes'
import { StructuralVariant } from './StructuralVariantPage'
import { textOrMissingTextWarning } from '../missingContent'

const FILTER_LABELS = {
  LOW_CALL_RATE: 'Low Call Rate',
  PCRPLUS_ENRICHED: 'PCR+ Enriched',
  UNRESOLVED: 'Unresolved',
  UNSTABLE_AF_PCRMINUS: 'Unstable AF PCR-',
  OUTLIER_SAMPLE_ENRICHED: 'Outlier Sample Enriched',
  LOWQUAL_WHAM_SR_DEL: 'Wham And Split-Read Evidence Only',

  IGH_MCH_OVERLAP: 'Overlaps Somatic Recombination Site',
  FAIL_MANUAL_REVIEW: 'Failed Manual Review',
  HIGH_NCR: 'High No-Call Rate',
}

const FILTER_DESCRIPTIONS = {
  LOW_CALL_RATE:
    'Site does not meet minimum requirements for fraction of PCR- samples with non-null genotypes. Flags sites more prone to false discoveries.',
  PCRPLUS_ENRICHED:
    'Site enriched for non-reference genotypes among PCR+ samples. Likely reflects technical batch effects. All PCR- samples have been assigned null GTs for these sites.',
  UNRESOLVED: 'Variant is unresolved',
  UNSTABLE_AF_PCRMINUS:
    'Allele frequency for this variant in PCR- samples is sensitive to choice of GQ filtering thresholds. All PCR- samples have been assigned null GTs for these sites.',
  OUTLIER_SAMPLE_ENRICHED:
    'SV was enriched for non-reference genotypes in outlier samples prior to filtering; may indicate a noisier or less reliable locus.',
  LOWQUAL_WHAM_SR_DEL:
    'SV was detected exclusively by Wham and only has split-read evidence, which usually correlates with lower-quality variants with elevated false discovery rates.',
  IGH_MCH_OVERLAP:
    'SV overlaps sites of known somatic recombination, such as immunoglobulin gene and/or major hisocompatability complex loci, making germline SV detecting and genotyping especially difficult.',
  FAIL_MANUAL_REVIEW: 'Variant failed manual review of evidence by a gnomAD analyst.',
  HIGH_NCR: 'Variant has excessively high rate of no-call (i.e., missing) genotypes.',
}

const EVIDENCE_LABELS: Record<string, string> = {
  BAF: 'Normalized B-allele frequency',
  PE: 'Anomalous paired-end reads',
  RD: 'Read depth',
  SR: 'Split reads',
}

const ALGORITHM_LABELS = {
  delly: 'DELLY',
  depth: 'Depth',
  manta: 'Manta',
  melt: 'MELT',
  wham: 'Wham',
  depth_manual: 'Manual curation in known genomic disorder region',
}

const COMPLEX_TYPE_LABELS = {
  CCR: 'Complex chromosomal rearrangement',
  dDUP: 'Dispersed duplication',
  dDUP_iDEL: 'Dispersed duplication with insert site deletion',
  delINV: 'Deletion-flanked inversion',
  delINVdel: 'Paired-deletion inversion',
  delINVdup: 'Inversion with flanking deletion and duplication',
  dupINV: 'Duplication-flanked inversion',
  dupINVdup: 'Paired-duplication inversion',
  dupINVdel: 'Inversion with flanking duplication and deletion',
  INVdel: 'Deletion-flanked inversion',
  INVdup: 'Duplication-flanked inversion',
  INS_iDEL: 'Insertion with insert site deletion',
  piDUP_FR: 'Palindromic inverted duplication',
  piDUP_RF: 'Palindromic inverted duplication',
  'CTX_PQ/QP': 'Reciprocal translocation of P:Q and Q:P arms',
  CTX_INV: 'Reciprocal translocation with derivative junction inversion',
  'CTX_PP/QQ': 'Reciprocal translocation of P:P and Q:Q arms',
}

const algorithmLabel = (algorithm: string) =>
  textOrMissingTextWarning('algorithm label', ALGORITHM_LABELS, algorithm)

const complexTypeLabel = (complexType: string) =>
  textOrMissingTextWarning('complex type label', COMPLEX_TYPE_LABELS, complexType)

const filterLabel = (filter: string) =>
  textOrMissingTextWarning('filter label', FILTER_LABELS, filter)

const filterDescription = (filter: string) =>
  textOrMissingTextWarning('filter description', FILTER_DESCRIPTIONS, filter)

type OwnPointLinkProps = {
  chrom: string
  pos: number
  windowSize?: number
}

// @ts-expect-error TS(2456) FIXME: Type alias 'PointLinkProps' circularly references ... Remove this comment to see the full error message
type PointLinkProps = OwnPointLinkProps & typeof PointLink.defaultProps

// @ts-expect-error TS(7022) FIXME: 'PointLink' implicitly has type 'any' because it d... Remove this comment to see the full error message
const PointLink = ({ chrom, pos, windowSize }: PointLinkProps) => (
  <Link to={`/region/${chrom}-${Math.max(pos - windowSize / 2, 0)}-${pos + windowSize / 2}`}>
    {chrom}:{pos}
  </Link>
)

PointLink.defaultProps = {
  windowSize: 10000,
}

type OwnComplexTypeHelpButtonProps = {
  complexType?: string
}

// @ts-expect-error TS(2456) FIXME: Type alias 'ComplexTypeHelpButtonProps' circularly... Remove this comment to see the full error message
type ComplexTypeHelpButtonProps = OwnComplexTypeHelpButtonProps &
  typeof ComplexTypeHelpButton.defaultProps

// @ts-expect-error TS(7022) FIXME: 'ComplexTypeHelpButton' implicitly has type 'any' ... Remove this comment to see the full error message
const ComplexTypeHelpButton = ({ complexType }: ComplexTypeHelpButtonProps) => {
  if (!complexType) {
    return null
  }

  // @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
  const helpTopic = {
    CCR: null,
    dDUP: 'sv-class_CPX_dDUP_dDUP-iDEL',
    dDUP_iDEL: 'sv-class_CPX_dDUP_dDUP-iDEL',
    delINV: 'sv-class_CPX_delINV_INVdel',
    delINVdel: 'sv-class_CPX_delINVdel',
    delINVdup: 'sv-class_CPX_delINVdup_dupINVdel',
    dupINV: 'sv-class_CPX_dupINV_INVdup',
    dupINVdel: 'sv-class_CPX_delINVdup_dupINVdel',
    dupINVdup: 'sv-class_CPX_dupINVdup',
    INS_iDEL: 'sv-class_CPX_INS-iDEL',
    INVdel: 'sv-class_CPX_delINV_INVdel',
    INVdup: 'sv-class_CPX_dupINV_INVdup',
    piDUP_FR: 'sv-class_CPX_piDUP-FR_piDUP-RF',
    piDUP_RF: 'sv-class_CPX_piDUP-FR_piDUP-RF',
  }[complexType]

  if (!helpTopic) {
    return null
  }

  return <InfoButton topic={helpTopic} />
}

ComplexTypeHelpButton.defaultProps = {
  complexType: null,
}

type StructuralVariantAttributeListProps = {
  variant: StructuralVariant
}

const StructuralVariantAttributeList = ({ variant }: StructuralVariantAttributeListProps) => {
  const filters = variant.filters || []
  const algorithms = variant.algorithms || []
  const length = variant.length || variant.end - variant.pos
  const evidence = variant.evidence || []
  const cpx_intervals = variant.cpx_intervals || []

  return (
    <AttributeList style={{ marginTop: '1.25em' }}>
      <AttributeListItem label="Filter">
        {filters.length > 0 ? (
          filters.map((filter) => (
            <Badge key={filter} level="warning" tooltip={filterDescription(filter)}>
              {filterLabel(filter)}
            </Badge>
          ))
        ) : (
          <Badge level="success">Pass</Badge>
        )}
      </AttributeListItem>
      <AttributeListItem label={variant.type === 'MCNV' ? 'Non-Diploid Samples' : 'Allele Count'}>
        {variant.ac}
      </AttributeListItem>
      <AttributeListItem label={variant.type === 'MCNV' ? 'Total Samples' : 'Allele Number'}>
        {variant.an}
      </AttributeListItem>
      <AttributeListItem
        label={variant.type === 'MCNV' ? 'Non-diploid CN Frequency' : 'Allele Frequency'}
      >
        {(variant.an === 0 ? 0 : variant.ac / variant.an).toPrecision(4)}
      </AttributeListItem>
      <AttributeListItem label="Quality score">{variant.qual}</AttributeListItem>
      <AttributeListItem label="Position">
        {variant.type === 'BND' || variant.type === 'CTX' || variant.type === 'INS' ? (
          <PointLink chrom={variant.chrom} pos={variant.pos} />
        ) : (
          <Link to={`/region/${variant.chrom}-${variant.pos}-${variant.end}`}>
            {variant.chrom}:{variant.pos}-{variant.end}
          </Link>
        )}
      </AttributeListItem>
      {(variant.type === 'BND' || variant.type === 'CTX') && (
        <AttributeListItem label="Second Position">
          <PointLink chrom={variant.chrom2} pos={variant.pos2} />
        </AttributeListItem>
      )}
      {variant.type !== 'BND' && variant.type !== 'CTX' && (
        <AttributeListItem label="Size">
          {length === -1 ? 'â€”' : `${length.toLocaleString()} bp`}
        </AttributeListItem>
      )}
      {variant.type && (
        <AttributeListItem label="Class">
          {svTypeLabels[variant.type]}{' '}
          {variant.type === 'INS' &&
            variant.alts &&
            `(${variant.alts
              .map((alt) => alt.replace(/^</, '').replace(/>$/, '').replace(/^INS:/, ''))
              .join(', ')})`}
          <InfoButton topic={`sv-class_${variant.type}`} />
        </AttributeListItem>
      )}
      {variant.type === 'CPX' && variant.cpx_type && (
        <React.Fragment>
          <AttributeListItem label="Complex SV Class">
            {variant.cpx_type} ({complexTypeLabel(variant.cpx_type)}){' '}
            <ComplexTypeHelpButton complexType={variant.cpx_type} />
          </AttributeListItem>
          <AttributeListItem label="Rearranged Segments">
            {cpx_intervals.join(', ')}
          </AttributeListItem>
        </React.Fragment>
      )}
      <AttributeListItem label="Evidence">
        {evidence.map((e) => EVIDENCE_LABELS[e]).join(', ')}
      </AttributeListItem>
      <AttributeListItem label="Algorithms">
        {algorithms.map((a) => algorithmLabel(a)).join(', ')}
      </AttributeListItem>
    </AttributeList>
  )
}

export default StructuralVariantAttributeList
