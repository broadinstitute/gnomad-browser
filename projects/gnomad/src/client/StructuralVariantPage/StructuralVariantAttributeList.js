import PropTypes from 'prop-types'
import React from 'react'

import { QuestionMark } from '@broad/help'
import { Badge } from '@broad/ui'

import AttributeList from '../AttributeList'
import Link from '../Link'
import { svTypeLabels } from '../StructuralVariantList/structuralVariantTypes'
import StructuralVariantDetailPropType from './StructuralVariantDetailPropType'

const FILTER_LABELS = {
  LOW_CALL_RATE: 'Low Call Rate',
  PCRPLUS_ENRICHED: 'PCR+ Enriched',
  UNRESOLVED: 'Unresolved',
  UNSTABLE_AF_PCRMINUS: 'Unstable AF PCR-',
}

const FILTER_DESCRIPTIONS = {
  LOW_CALL_RATE:
    'Site does not meet minimum requirements for fraction of PCR- samples with non-null genotypes. Flags sites more prone to false discoveries.',
  PCRPLUS_ENRICHED:
    'Site enriched for non-reference genotypes among PCR+ samples. Likely reflects technical batch effects. All PCR- samples have been assigned null GTs for these sites.',
  UNRESOLVED: 'Variant is unresolved',
  UNSTABLE_AF_PCRMINUS:
    'Allele frequency for this variant in PCR- samples is sensitive to choice of GQ filtering thresholds. All PCR- samples have been assigned null GTs for these sites.',
}

const EVIDENCE_LABELS = {
  BAF: 'Normalized B-allele frequency',
  PE: 'Anomalous paired-end reads',
  RD: 'Read depth',
  SR: 'Split reads',
}

const ALGORITHM_LABELS = {
  delly: 'DELLY',
  depth: 'Depth',
  manta: 'Manta',
  melt: 'MELT',
}

const COMPLEX_TYPE_LABELS = {
  CCR: 'Complex chromosomal rearrangement',
  dDUP: 'Dispersed duplication',
  dDUP_iDEL: 'Dispersed duplication with insert site deletion',
  delINV: 'Deletion-flanked inversion',
  delINVdel: 'Paired-deletion inversion',
  delINVdup: 'Inversion with flanking deletion and duplication',
  dupINV: 'Duplication-flanked inversion',
  dupINVdup: 'Paired-duplication inversion',
  dupINVdel: 'Inversion with flanking duplication and deletion',
  INVdel: 'Deletion-flanked inversion',
  INVdup: 'Duplication-flanked inversion',
  INS_iDEL: 'Insertion with insert site deletion',
  piDUP_FR: 'Palindromic inverted duplication',
  piDUP_RF: 'Palindromic inverted duplication',
}

const PointLink = ({ chrom, pos, windowSize }) => (
  <Link to={`/region/${chrom}-${Math.max(pos - windowSize / 2, 0)}-${pos + windowSize / 2}`}>
    {chrom}:{pos}
  </Link>
)

PointLink.propTypes = {
  chrom: PropTypes.string.isRequired,
  pos: PropTypes.number.isRequired,
  windowSize: PropTypes.number,
}

PointLink.defaultProps = {
  windowSize: 10000,
}

const ComplexTypeHelpButton = ({ complexType }) => {
  if (!complexType) {
    return null
  }

  const helpTopic = {
    CCR: null,
    dDUP: 'sv-class_CPX_dDUP_dDUP-iDEL',
    dDUP_iDEL: 'sv-class_CPX_dDUP_dDUP-iDEL',
    delINV: 'sv-class_CPX_delINV_INVdel',
    delINVdel: 'sv-class_CPX_delINVdel',
    delINVdup: 'sv-class_CPX_delINVdup_dupINVdel',
    dupINV: 'sv-class_CPX_dupINV_INVdup',
    dupINVdel: 'sv-class_CPX_delINVdup_dupINVdel',
    dupINVdup: 'sv-class_CPX_dupINVdup',
    INS_iDEL: 'sv-class_CPX_INS-iDEL',
    INVdel: 'sv-class_CPX_delINV_INVdel',
    INVdup: 'sv-class_CPX_dupINV_INVdup',
    piDUP_FR: 'sv-class_CPX_piDUP-FR_piDUP-RF',
    piDUP_RF: 'sv-class_CPX_piDUP-FR_piDUP-RF',
  }[complexType]

  if (!helpTopic) {
    return null
  }

  return <QuestionMark topic={`SV_docs/${helpTopic}`} />
}

ComplexTypeHelpButton.propTypes = {
  complexType: PropTypes.string,
}

ComplexTypeHelpButton.defaultProps = {
  complexType: null,
}

const StructuralVariantAttributeList = ({ variant }) => (
  <AttributeList labelWidth={variant.type === 'MCNV' ? 180 : 140} style={{ marginTop: '1.25em' }}>
    <AttributeList.Item label="Filter">
      {variant.filters.length > 0 ? (
        variant.filters.map(filter => (
          <Badge key={filter} level="warning" tooltip={FILTER_DESCRIPTIONS[filter]}>
            {FILTER_LABELS[filter] || filter}
          </Badge>
        ))
      ) : (
        <Badge level="success">Pass</Badge>
      )}
    </AttributeList.Item>
    <AttributeList.Item label={variant.type === 'MCNV' ? 'Non-Diploid Samples' : 'Allele Count'}>
      {variant.ac}
    </AttributeList.Item>
    <AttributeList.Item label={variant.type === 'MCNV' ? 'Total Samples' : 'Allele Number'}>
      {variant.an}
    </AttributeList.Item>
    <AttributeList.Item
      label={variant.type === 'MCNV' ? 'Non-diploid CN Frequency' : 'Allele Frequency'}
    >
      {(variant.an === 0 ? 0 : variant.ac / variant.an).toPrecision(4)}
    </AttributeList.Item>
    <AttributeList.Item label="Quality score">{variant.qual}</AttributeList.Item>
    <AttributeList.Item label="Position">
      {variant.type === 'BND' || variant.type === 'CTX' || variant.type === 'INS' ? (
        <PointLink chrom={variant.chrom} pos={variant.pos} />
      ) : (
        <Link to={`/region/${variant.chrom}-${variant.pos}-${variant.end}`}>
          {variant.chrom}:{variant.pos}-{variant.end}
        </Link>
      )}
    </AttributeList.Item>
    {(variant.type === 'BND' || variant.type === 'CTX') && (
      <AttributeList.Item label="Second Position">
        <PointLink chrom={variant.chrom2} pos={variant.pos2} />
      </AttributeList.Item>
    )}
    {variant.type !== 'BND' && variant.type !== 'CTX' && (
      <AttributeList.Item label="Size">
        {variant.length === -1 ? 'â€”' : `${variant.length.toLocaleString()} bp`}
      </AttributeList.Item>
    )}
    <AttributeList.Item label="Class">
      {svTypeLabels[variant.type]}{' '}
      {variant.type === 'INS' &&
        variant.alts &&
        `(${variant.alts
          .map(alt =>
            alt
              .replace(/^</, '')
              .replace(/>$/, '')
              .replace(/^INS:/, '')
          )
          .join(', ')})`}
      <QuestionMark topic={`SV_docs/sv-class_${variant.type}`} />
    </AttributeList.Item>
    {variant.type === 'CPX' && (
      <React.Fragment>
        <AttributeList.Item label="Complex SV Class">
          {variant.cpx_type} ({COMPLEX_TYPE_LABELS[variant.cpx_type]}){' '}
          <ComplexTypeHelpButton complexType={variant.cpx_type} />
        </AttributeList.Item>
        <AttributeList.Item label="Rearranged Segments">
          {variant.cpx_intervals.join(', ')}
        </AttributeList.Item>
      </React.Fragment>
    )}
    <AttributeList.Item label="Evidence">
      {variant.evidence.map(e => EVIDENCE_LABELS[e]).join(', ')}
    </AttributeList.Item>
    <AttributeList.Item label="Algorithms">
      {variant.algorithms.map(a => ALGORITHM_LABELS[a]).join(', ')}
    </AttributeList.Item>
  </AttributeList>
)

StructuralVariantAttributeList.propTypes = {
  variant: StructuralVariantDetailPropType.isRequired,
}

export default StructuralVariantAttributeList
