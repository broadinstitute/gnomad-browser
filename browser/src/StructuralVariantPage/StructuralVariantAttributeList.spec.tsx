import { expect, test } from '@jest/globals'
import 'jest-styled-components'

import React from 'react'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

import StructuralVariantAttributeList from './StructuralVariantAttributeList'
import StructuralVariantFactory from '../__factories__/StructuralVariant'

test('StructuralVariantAttributeList tolerates certain variant fields being null', () => {
  const variant = StructuralVariantFactory.build({
    filters: null,
    algorithms: null,
    length: null,
    evidence: null,
    cpx_intervals: null,
  })
  const tree = renderer.create(
    <BrowserRouter>
      <StructuralVariantAttributeList variant={variant} />
    </BrowserRouter>
  )
  expect(tree).toMatchSnapshot()
})
