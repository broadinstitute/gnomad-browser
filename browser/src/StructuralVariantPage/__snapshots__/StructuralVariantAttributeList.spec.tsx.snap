// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructuralVariantAttributeList tolerates certain variant fields being null 1`] = `
.c1 {
  position: relative;
  top: -0.1em;
  display: inline-block;
  padding: 0.25em 0.4em 0.2em;
  border: 1px solid #000;
  border-radius: 0.3em;
  background: #2E7D32;
  color: #fff;
  font-size: 0.75em;
  font-weight: bold;
  line-height: 1;
}

.c0 {
  margin: 0;
}

.c0 dt,
.c0 dd {
  display: inline-block;
  line-height: 1.75;
}

.c0 dt {
  font-weight: bold;
  vertical-align: top;
}

.c0 dd {
  margin-left: 0.5ch;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  outline: none;
  padding: 0 3px;
  border: none;
  background: none;
  cursor: pointer;
}

.c3 img {
  position: relative;
  top: 0.13em;
  width: 14px;
  height: 14px;
  border-radius: 7px;
}

.c3:focus img {
  box-shadow: 0 0 0 0.2em rgba(70,130,180,0.5);
}

.c2 {
  color: #1173bb;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:visited,
.c2:active {
  color: #1173bb;
}

.c2:focus,
.c2:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

@media (max-width:600px) {
  .c0 dt,
  .c0 dd {
    display: block;
  }

  .c0 dd {
    margin-left: 2ch;
  }
}

<dl
  className="c0"
  style={
    {
      "marginTop": "1.25em",
    }
  }
>
  <div>
    <dt>
      Filter
    </dt>
    <dd>
      <span
        className="c1"
      >
        Pass
      </span>
    </dd>
  </div>
  <div>
    <dt>
      Allele Count
    </dt>
    <dd>
      123
    </dd>
  </div>
  <div>
    <dt>
      Allele Number
    </dt>
    <dd>
      345
    </dd>
  </div>
  <div>
    <dt>
      Allele Frequency
    </dt>
    <dd>
      0.3565
    </dd>
  </div>
  <div>
    <dt>
      Quality score
    </dt>
    <dd>
      99
    </dd>
  </div>
  <div>
    <dt>
      Position
    </dt>
    <dd>
      <a
        className="-Link c2"
        href="/region/21-123-456"
        onClick={[Function]}
      >
        21
        :
        123
        -
        456
      </a>
    </dd>
  </div>
  <div>
    <dt>
      Size
    </dt>
    <dd>
      333 bp
    </dd>
  </div>
  <div>
    <dt>
      Class
    </dt>
    <dd>
      duplication
       
      <button
        className="c3"
        onClick={[Function]}
        type="button"
      >
        <img
          alt=""
          aria-hidden="true"
          src="test-file-stub"
        />
        <span
          style={
            {
              "border": "0",
              "clip": "rect(0 0 0 0)",
              "height": "1px",
              "margin": "-1px",
              "overflow": "hidden",
              "padding": "0",
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "1px",
            }
          }
        >
          More information
        </span>
      </button>
    </dd>
  </div>
  <div>
    <dt>
      Evidence
    </dt>
    <dd>
      
    </dd>
  </div>
  <div>
    <dt>
      Algorithms
    </dt>
    <dd>
      
    </dd>
  </div>
</dl>
`;
